interp/src/primitives.rs

```diff
@@ -1,224 +1,1846 @@
+            inputs.iter().find(|(id, _)| id == "write_data").unwrap();
+        let (_, write_en) =
+            inputs.iter().find(|(id, _)| id == "write_en").unwrap();
+        let (_, addr0) = inputs.iter().find(|(id, _)| id == "addr0").unwrap();
+        //check that addr0 is not out of bounds and that it is the proper width!
+        check_widths(addr0, addr0, self.idx_size); //make a unary one instead of hacking. Also change the panicking
+                                                   //so we don't have to keep using .as_u64()
+        let addr0 = addr0.as_u64();
+        if addr0 >= self.size {
+            panic!(
+                "memory only has {} slots, addr0 tries to access slot {}",
+                self.size, addr0
+            );
+        }
+        //check that input data is the appropriate width as well
+        check_width(input, self.width);
+        let old = self.data[addr0 as usize].clone();
+        // only write to memory if write_en is 1
+        if write_en.as_u64() == 1 {
+            self.data[addr0 as usize] = (*input).clone();
+            // what's in this vector:
+            // the "out" -- TimeLockedValue ofthe new mem data. Needs 1 cycle before readable
+            // "done" -- TimeLockedValue of DONE, which is asserted 1 cycle after we write
+            // all this coordination is done by the interpreter. We just set it up correctly
+            vec![
+                (
+                    ir::Id::from("read_data"),
+                    TimeLockedValue::new(
+                        self.data[addr0 as usize].clone(),
+                        1,
+                        Some(old),
+                    )
+                    .into(),
+                ),
+                ("done".into(), PulseValue::one_cycle_one_bit_pulse().into()),
+            ]
+        } else {
+            // if write_en was low, so done is 0 b/c nothing was written here
+            // in this vector i
+            // READ_DATA: (immediate), just the old value b/c write was unsuccessful
+            // DONE: not TimeLockedValue, b/c it's just 0, b/c our write was unsuccessful
+            vec![
+                (ir::Id::from("read_data"), old.into()),
+                // (ir::Id::from("done"), Value::zeroes(1).into()),
+            ]
+        }
+    }
+
+    fn reset(&self, inputs: &[(ir::Id, &Value)]) -> Vec<(ir::Id, OutputValue)> {
+        let (_, addr0) = inputs.iter().find(|(id, _)| id == "addr0").unwrap();
+        //check that addr0 is not out of bounds and that it is the proper width!
+        check_widths(addr0, addr0, self.idx_size); //make a unary one instead of hacking. Also change the panicking
+                                                   //so we don't have to keep using .as_u64()
+        let addr0 = addr0.as_u64();
+        if addr0 >= self.size {
+            panic!(
+                "memory only has {} slots, addr0 tries to access slot {}",
+                self.size, addr0
+            );
+        }
```

Comment:
This `check_width(...), if addr0 >= self.size` pattern repeats a lot of times in this file. Make this a function/method on the impls?

Reply:
refactored (painfully) into a trait